instrumentation/armeria-1.0/library/src/main/java/io/opentelemetry/instrumentation/armeria/v1_0/client/OpenTelemetryClient.java

```diff
@@ -0,0 +1,103 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.instrumentation.armeria.v1_0.client;
+
+import com.linecorp.armeria.client.ClientRequestContext;
+import com.linecorp.armeria.client.HttpClient;
+import com.linecorp.armeria.client.SimpleDecoratingHttpClient;
+import com.linecorp.armeria.common.HttpRequest;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.logging.RequestLogProperty;
+import io.opentelemetry.auto.bootstrap.instrumentation.decorator.BaseTracer;
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.trace.Span;
+import io.opentelemetry.trace.Tracer;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+
+/** Decorates an {@link HttpClient} to trace outbound {@link HttpResponse}s. */
+public class OpenTelemetryClient extends SimpleDecoratingHttpClient {
+
+  /** Creates a new tracing {@link HttpClient} decorator using the default {@link Tracer}. */
+  public static OpenTelemetryClient.Decorator newDecorator() {
+    return new Decorator(new ArmeriaClientTracer());
+  }
+
+  /** Creates a new tracing {@link HttpClient} decorator using the specified {@link Tracer}. */
+  public static OpenTelemetryClient.Decorator newDecorator(Tracer tracer) {
+    return new Decorator(new ArmeriaClientTracer(tracer));
+  }
+
+  /**
+   * Creates a new tracing {@link HttpClient} decorator using the specified {@link
+   * ArmeriaClientTracer}.
+   */
+  public static OpenTelemetryClient.Decorator newDecorator(ArmeriaClientTracer clientTracer) {
+    return new Decorator(clientTracer);
+  }
+
+  private final ArmeriaClientTracer clientTracer;
+
+  private OpenTelemetryClient(HttpClient delegate, ArmeriaClientTracer clientTracer) {
+    super(delegate);
+    this.clientTracer = clientTracer;
+  }
+
+  @Override
+  public HttpResponse execute(ClientRequestContext ctx, HttpRequest req) throws Exception {
+    // Always available in practice.
+    long requestStartTimeMicros =
+        ctx.log().ensureAvailable(RequestLogProperty.REQUEST_START_TIME).requestStartTimeMicros();
+    long requestStartTimeNanos = TimeUnit.MICROSECONDS.toNanos(requestStartTimeMicros);
+    Span span = clientTracer.startSpan(ctx, requestStartTimeNanos);
+
+    if (span.isRecording()) {
+      ctx.log()
+          .whenComplete()
+          .thenAccept(
+              log -> {
+                BaseTracer.onPeerConnection(span, ctx.remoteAddress());
+
+                long requestEndTimeNanos = requestStartTimeNanos + log.responseDurationNanos();
+                if (log.responseCause() != null) {
+                  clientTracer.endExceptionally(
+                      span, log, log.responseCause(), requestEndTimeNanos);
+                } else {
+                  clientTracer.end(span, log, requestEndTimeNanos);
+                }
+              });
+    }
+
+    try (Scope ignored = clientTracer.startScope(span, ctx)) {
```

Comment:
Just to be sure: you start new scope even if span is not recording?

Reply:
Yeah - I think it's required for logs injection

Reply:
ah, interesting, hadn't thought of that 👍